### ISSUE TITLE
"Maximum function nesting level of '100' reached" on low number on parallel retries

### ISSUE THREAD

**Issue by paslandau** [guzzle/guzzle#964]
This issue might be related to https://github.com/guzzle/guzzle/issues/905

I noticed that Guzzle will terminated with a "Maximum function nesting level of '100' reached" fatal error when some retries (via `AbstractRetryableEvent::retry()`) are made in parallel. Code to reproduce:

```php
$requestNum = 5;
$retryPerRequest = 5;

$c = new Client();

/* Bug is unfortunately not reproducable when using the Mock Subscriber */
//$responses = [];
//for($i = 0; $i < $requestNum*$retryPerRequest; $i++){
//    $responses[] = new Response(200);
//}
//$mock = new Mock($responses);
//$c->getEmitter()->attach($mock);

$rq = [];
for($i = 0; $i<$requestNum; $i++){
    $request = $c->createRequest("GET","http://example.org/");
    $request->getConfig()->set("id",$i);
    $rq[] = $request;
}

$i = 0;
$f = function(AbstractRetryableEvent $event)  use ($retryPerRequest){
    $retries = $event->getRequest()->getConfig()->get("retries");
    if($retries === null){
        $retries = 0;
    }
    if($retries < $retryPerRequest){
        $retries++;
        $event->getRequest()->getConfig()->set("retries",$retries);
        $id = $event->getRequest()->getConfig()->get("id");
        echo "Request $id, Retry $retries\n";
        $event->retry();
    }
};
$o = [
    "pool_size" => 25,
    "error" => $f,
    "complete" => $f,
    "end" => function(EndEvent $event) use (&$i){
        $id = $event->getRequest()->getConfig()->get("id");
        echo "Finished $id\n";
    }
];
$pool = new Pool($c,$rq,$o);
$pool->wait();
```

Result
```
Request 0, Retry 1
Request 1, Retry 1
Request 2, Retry 1
Request 3, Retry 1
Request 4, Retry 1
Request 0, Retry 2
Fatal error: Maximum function nesting level of '100' reached, aborting! in D:\xampp\htdocs\paslandau\vendor\guzzlehttp\guzzle\src\RequestFsm.php on line 146

Call Stack:
    0.0000     135408   1. {main}() D:\xampp\htdocs\paslandau\tests\guzzle-mass-retry-test.php:0
    0.0190    1177184   2. GuzzleHttp\Pool->wait() D:\xampp\htdocs\paslandau\tests\guzzle-mass-retry-test.php:59
    0.0280    1406856   3. GuzzleHttp\Message\FutureResponse->wait() D:\xampp\htdocs\paslandau\vendor\guzzlehttp\guzzle\src\Pool.php:189
    0.0280    1409080   4. GuzzleHttp\Message\FutureResponse->invokeWait() D:\xampp\htdocs\paslandau\vendor\guzzlehttp\ringphp\src\Future\BaseFutureTrait.php:55
    0.0280    1409208   5. GuzzleHttp\Ring\Future\FutureArray->wait() D:\xampp\htdocs\paslandau\vendor\guzzlehttp\ringphp\src\Future\BaseFutureTrait.php:118
    0.0280    1411432   6. GuzzleHttp\Ring\Future\FutureArray->invokeWait() D:\xampp\htdocs\paslandau\vendor\guzzlehttp\ringphp\src\Future\BaseFutureTrait.php:55
    0.0280    1411592   7. GuzzleHttp\Ring\Client\CurlMultiHandler->execute() D:\xampp\htdocs\paslandau\vendor\guzzlehttp\ringphp\src\Future\BaseFutureTrait.php:118
    0.2630    1413192   8. GuzzleHttp\Ring\Client\CurlMultiHandler->processMessages() D:\xampp\htdocs\paslandau\vendor\guzzlehttp\ringphp\src\Client\CurlMultiHandler.php:125
    0.2640    1420504   9. React\Promise\Deferred->resolve() D:\xampp\htdocs\paslandau\vendor\guzzlehttp\ringphp\src\Client\CurlMultiHandler.php:236
    0.2640    1420536  10. call_user_func() D:\xampp\htdocs\paslandau\vendor\react\promise\src\Deferred.php:35
    0.2640    1420568  11. React\Promise\Promise->React\Promise\{closure}() D:\xampp\htdocs\paslandau\vendor\react\promise\src\Deferred.php:35
    0.2640    1420600  12. React\Promise\Promise->resolve() D:\xampp\htdocs\paslandau\vendor\react\promise\src\Promise.php:169
    0.2640    1433064  13. React\Promise\Promise->settle() D:\xampp\htdocs\paslandau\vendor\react\promise\src\Promise.php:128
    0.2640    1433088  14. React\Promise\Promise->React\Promise\{closure}() D:\xampp\htdocs\paslandau\vendor\react\promise\src\Promise.php:156
    0.2640    1433120  15. React\Promise\FulfilledPromise->then() D:\xampp\htdocs\paslandau\vendor\react\promise\src\Promise.php:114
[...]
    0.8230    1679744  95. React\Promise\Promise->React\Promise\{closure}() D:\xampp\htdocs\paslandau\vendor\react\promise\src\Deferred.php:35
    0.8230    1679744  96. React\Promise\Promise->resolve() D:\xampp\htdocs\paslandau\vendor\react\promise\src\Promise.php:169
    0.8230    1679816  97. React\Promise\Promise->settle() D:\xampp\htdocs\paslandau\vendor\react\promise\src\Promise.php:128
    0.8230    1679840  98. React\Promise\Promise->React\Promise\{closure}() D:\xampp\htdocs\paslandau\vendor\react\promise\src\Promise.php:156
    0.8230    1679840  99. React\Promise\FulfilledPromise->then() D:\xampp\htdocs\paslandau\vendor\react\promise\src\Promise.php:114
```

5 Requests are made in parallel. Every request ist retried immediately in either the `complete` or `error` event for a maximum of 5 times. I'm not sure what makes the stack grow so big, but as you can see only ~6 requests are being made until the error occurs.

You can also adjust the code snippet to use only 1 request and 10 retries to get the same result. ($requestNum = 1; $retryPerRequest = 10;)

I was unable to reproduce the behaviour using the `Mock` Subscriber so I had to make "real" requests against some website.

Any idea what's causing this and how to prevent it?

Cheers
Pascal

**Comment by mtdowling** [comment]
Request retries currently use recursion. I'm open to changing this if possible (though I'm struggling to come up with a non-recursive solution). You'll need to increase your stack frame size limitation that is imposed by XDebug (100 is really low for any and usecase language).

**Comment by mtdowling** [comment]
Here's a workaround for now until a better solution is found (either by addressing the recursion I linked to or addressing the recursion in the react/promises library): https://github.com/guzzle/guzzle/pull/967. Can you let me know if that fixes the issue for you?

**Comment by mtdowling** [comment]
I think this was fixed by #967. Let me know otherwise.

**Comment by paslandau** [comment]
Hey,
sorry for the late reply. It fixes the original issue for me, thx for the quick solution!
